File: apps/tilawa/android/app/src/main/kotlin/com/tilawa/app/widget/athkar/AthkarWidgetPayload.kt
package com.tilawa.app.widget.athkar

import org.json.JSONArray
import org.json.JSONObject

internal data class AthkarWidgetItem(
    val text: String,
    val count: Int,
)

/**
 * Parsed payload of the athkar widget snapshot (spec 041, US3). Carries both
 * sets; the provider picks the applicable window by local time.
 */
internal data class AthkarWidgetPayload(
    val morningTitle: String,
    val eveningTitle: String,
    val morning: List<AthkarWidgetItem>,
    val evening: List<AthkarWidgetItem>,
) {
    companion object {
        /** Returns null when either set is missing or empty. */
        fun parse(payload: JSONObject): AthkarWidgetPayload? {
            val morning = parseItems(payload.optJSONArray("morning"))
            val evening = parseItems(payload.optJSONArray("evening"))
            if (morning.isEmpty() || evening.isEmpty()) return null
            return AthkarWidgetPayload(
                morningTitle = payload.optString("morningTitle"),
                eveningTitle = payload.optString("eveningTitle"),
                morning = morning,
                evening = evening,
            )
        }

        private fun parseItems(array: JSONArray?): List<AthkarWidgetItem> {
            if (array == null) return emptyList()
            val items = ArrayList<AthkarWidgetItem>(array.length())
            for (i in 0 until array.length()) {
                val item = array.optJSONObject(i) ?: continue
                val text = item.optString("text")
                if (text.isBlank()) continue
                items.add(
                    AthkarWidgetItem(
                        text = text,
                        count = item.optInt("count", 1).coerceAtLeast(1),
                    ),
                )
            }
            return items
        }
    }
}

File: apps/tilawa/lib/features/islamic_widgets/domain/entities/athkar_widget_payload.dart
/// One dhikr entry displayed by the athkar widget.
class AthkarWidgetItem {
  const AthkarWidgetItem({required this.text, required this.count});

  /// Arabic dhikr text (athkar content is Arabic-only by design).
  final String text;

  /// Prescribed repetition count (display badge; the widget advances per
  /// item, not per repetition, in v1).
  final int count;

  Map<String, Object?> toJson() => <String, Object?>{
    'text': text,
    'count': count,
  };
}

/// Display-ready content for the morning/evening athkar widget (spec 041,
/// US3). Both sets ship in one snapshot; the native host picks the applicable
/// window by local time so no Dart isolate is needed at flip time.
class AthkarWidgetPayload {
  const AthkarWidgetPayload({
    required this.morningTitle,
    required this.eveningTitle,
    required this.morning,
    required this.evening,
  });

  final String morningTitle;
  final String eveningTitle;
  final List<AthkarWidgetItem> morning;
  final List<AthkarWidgetItem> evening;

  Map<String, Object?> toJson() => <String, Object?>{
    'morningTitle': morningTitle,
    'eveningTitle': eveningTitle,
    'morning': <Object?>[for (final item in morning) item.toJson()],
    'evening': <Object?>[for (final item in evening) item.toJson()],
  };
}
